### DISCUSSION TITLE
Is prior predictive check possible ?

### DISCUSSION THREAD

**Discussion by osupplisson** [inlabru-org/inlabru#229] [Q&A] [answered]
Hi,

I finally found out some georeferenced data and decided to give inlabru an in-depth try: truly amazing work !

To ensure my priors are not too dumb, I would like to perform a 'prior predictive' check (https://mc-stan.org/docs/stan-users-guide/posterior-predictive-checks.html). Is there any hidden function for that in the package? 

I tried with the `bru_run = FALSE` option + the `predict()` function but it did not work. I haven't found anything else to try in bru (nor in INLA). Am I missing a 'hidden' function for sampling from the initial configuration (i.e., predict from the prior model)? 

Is that even theoretically possible with the R-INLA approach or is it only a matter of implementation?

Best,
Olivier

**Comment by finnlindgren** [comment, marked as answer]
There is unfortunately not yet any automated interface for prior sampling in the INLA package; I think it's mostly an issue of someone taking the time to write the needed code to make it automatic.

A proper approach could be to first sample from all the "hyperparameters", and then call `inla()` with those as known values but without any observation input (would likely require one single `NA` "observation" to appease the input "sanity checks"), specifying the simulated hyperparameters as fixed and known. Then a call to `inla.posterior.sample()` would sample from the conditional prior for the latent variables. Repeating the whole procedure for a new sample of hyperparameter values, one would then get multiple samples from the full prior model.

However, in general, one would need to decide how to handle improper priors; in particular intrinsic stationary models such as rw1 and rw2; those by default have sum-to-zero constraints, making them proper with respect to their average values, but the rw2 model is still invariant to linear slopes, so prior samples are not well defined. When `constr = FALSE`, they are both fundamentally improper. A "safe" implementation of prior sampling would need to detect those types of models and at the very least warn the user about them. Fortunately, such "rank-deficient models" (without sufficient constraints) do not generate sensible prior samples (as they do not have well defined sampling properties until conditioning on observations), so refusing to sample from them in this context wouldn't necessarily be a bad thing.

**Reply by osupplisson** [reply]
Thank you very much for your reply and the points raised in the second paragraph !

I hadn't though of the approach you are describing in §1, which may be worthwhile to try.

Best,
Olivier
